teps/0146-parameters-in-script.md

```diff
@@ -0,0 +1,605 @@
+> `Containers`, like what an `entrypoint` or a `command` is.
+>
+> Users should only need to have passing familiarity with a shell environment
+> to be successful using Tekton.
+
+Problems with variable subsitution was noted in [original proposal][pipeline-issue-781]:
+
+> How this syntax works with input substitutions, e.g., ``${inputs.params.flags}`` might
+> also lead to confusion among users, which we should design for.
+
+Users need a way to safely and reliably substitute `Parameters` in `Scripts` in `Steps`
+of their `Tasks`. This would make it easier to use Tekton without learning advanced topics,
+while still using it reliably and safely.
+
+### Requirements
+
+<!--
+Describe constraints on the solution that must be met, such as:
+- which performance characteristics that must be met?
+- which specific edge cases that must be handled?
+- which user scenarios that will be affected and must be accommodated?
+-->
+
+1. **[Secure by default in v1][secure-by-default]**: Out of the box
+   *Tekton Pipelines* v1 should disallow common anti-patterns that
+   can render a default install insecure.
+1. **Easy to use**: Whatever replaces the direct use of `$(params.X)` in
+   `script` should be unsurprising and memorable to ease friction in user
+   migration.
+1. **Automatable**: A script or program should be able to convert `Tasks`
+   to the new approach. This might be impossible for all cases (e.g. if an
+   unrecognized shebang has been used in the script) but a best-effort
+   implementation should be provided.
+1. **Backwards-compatible in v1**. If a user already has `Tasks`
+   that use `$(params.X)` syntax in `script` then we should continue supporting
+   that usage in v1 resources. Ideally, Tekton Pipelines would warn the user
+   about it somehow.
+1. **(To be discussed) Provide an escape hatch in v1**. If a cluster
+   operator or admin decides that they want to continue allowing
+   `Parameter` variables to be directly injectable to `script` fields of
+   `Steps` there should be some way for them to configure that.
+
+## Proposal - Implicitly Project `Parameters` as Environment Variables in `Steps` (Language-Agnostic)
+We have projected `Parameters` as environment variables in the `git-clone` Task in the
+Catalog and in our own plumbing repository. We can use this approach, and simplify the
+syntax by implicitly projecting `Parameters` as environment variables in `Steps`.
+The environment variable should be prefixed with `PARAM_` to make it clear and predictable.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone ${$(params.url)} # <--- needs to be manually updated w/ additional ${*} wrapping for shell usage
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone ${PARAM_URL} 
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_URL
+      value: $(params.url)
```

Comment:
An alternative to mutating the step specification could be to inject the env in the `Pod`.
We could do that by default for all params by default.

There are some gotchas:
- the translation from param name to env variable name risks being brittle
- what happens in case of conflicts with user-defined environment variables or how do we prevent such conflicts
- how do we handle array params? We added support for indexing into array params but how do we render the list into an environment variable that is usable by different kinds of script environments? Shall we use a JSON blob? That's not too bash-friendly but very python-friendly
- how do we handle object params? Do we add one variable for each field? Or a single variable with a JSON blob?

Reply:
The last 2 questions of @afrittoli are also valid independently of "mutating the step spec" or not, right ?

Reply:
+1 to the gotchas outlined, I will address each one inline:
* the translation from param name to env variable name risks being brittle
    * This is a good point, I have not added this into the document but to reduce brittless in the param-name- > env-var name translation we could add a param name hash suffix (similar to k8s pods) to avoid collisions eg: TEKTON_PARAM_FOO_X5GHA.  This would make directly knowing the name for the param a bit more difficult for users but Tekton is dynamically generating these anyway so it should be ok if it is somewhat opaque
* what happens in case of conflicts with user-defined environment variables or how do we prevent such conflicts
    * Good point, this case important to call out explicitly.  I’ve added this into the document now under “Considered Edge Cases and Gotchas”.  The gist is that in case of an overlap with user-defined environment variables, Tekton will throw a validation error stating that the param should be renamed. NOTE: collision should be very unlikely given the TEKTON_PARAM prefix and we can also optionally add a sha suffix to the env var (based on the param name) to further unique-ify the env var name to reduce collision.
* how do we handle array params? We added support for indexing into array params but how do we render the list into an environment variable that is usable by different kinds of script environments? Shall we use a JSON blob? That's not too bash-friendly but very python-friendly
    * Currently the plan would be that the values in the env var map 1:1 with how they are currently used as that UX is working for users atm.  In the case of an array param, it would be the same values based on the syntax, for an array param + index (params.arr[0]) we would get the string value and for the entire array we would provide the list as a comma separated array (params.arr -> first,second,third…)  The TEP suggestion is built around the `script` field which only supports string type variables which currently has comma separated serialization for the array and actually has a validation error against using an object directly (vs an object field).  I have added some examples under the “Suggested Solution”.
* how do we handle object params? Do we add one variable for each field? Or a single variable with a JSON blob?
    * There would be a variable for each field.  I have added some examples under the “Suggested Solution”.  Currently in my experimentation around this it seems that you get a validation error if you try to reference a param object in a script directly (vs a param object field which works).